### PR TITLE
chore(python): Start at using new Bound<> API from PyO3

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #15699

There will be more follow-up PRs, this is just a small chunk of the usages that need upgrading (although hopefully not as tricky).

### CODE REVIEW

File: py-polars/src/conversion/any_value.rs [hunk 1]
```diff
@@ -371,51 +390,50 @@ pub(crate) fn py_object_to_any_value(ob: &PyAny, strict: bool) -> PyResult<AnyVa
...
-        LUT.with_gil(py, |lut| {
-            let convert_fn = lut
-                .entry(type_object_ptr)
-                .or_insert_with(|| get_conversion_function(ob, py));
+    // let type_object_ptr = ob.get_type().as_type_ptr() as usize;
 
-            convert_fn(ob, strict)
-        })
-    })
+    // Python::with_gil(|py| {
```
Comment (itamarst) [inline py-polars/src/conversion/any_value.rs:432]: @ritchie46 I can't figure out how to make this compile, maybe you have some ideas?

File: py-polars/src/conversion/any_value.rs [hunk 2]
```diff
@@ -371,51 +390,50 @@ pub(crate) fn py_object_to_any_value(ob: &PyAny, strict: bool) -> PyResult<AnyVa
             match &*type_name {
                 // Can't use pyo3::types::PyDateTime with abi3-py37 feature,
                 // so need this workaround instead of `isinstance(ob, datetime)`.
-                "date" => get_date,
-                "time" => get_time,
-                "datetime" => get_datetime,
-                "timedelta" => get_timedelta,
-                "Decimal" => get_decimal,
-                "range" => get_list,
+                "date" => get_date as InitFn,
```
Comment (itamarst) [inline py-polars/src/conversion/any_value.rs:393]: Lacking the `as InitFn` I got errors about "expected fn pointer, got fn item" (or vice versa?).

File: py-polars/src/conversion/any_value.rs (hunk 1 above)
Comment (ritchie46) [reply py-polars/src/conversion/any_value.rs]: Yes. See comments.

File: py-polars/src/conversion/any_value.rs [hunk 3]
```diff
@@ -115,7 +115,7 @@ pub(crate) fn any_value_into_py_object(av: AnyValue, py: Python) -> PyObject {
 }
 
 type TypeObjectPtr = usize;
-type InitFn<'py> = fn(&Bound<'py, PyAny>, bool) -> PyResult<AnyValue<'py>>;
```
Comment (ritchie46) [inline py-polars/src/conversion/any_value.rs]: This made the lifetime fixed to `'py`. But given that we put `InitFn` in a global LUT, the materialization of `'py` is not bounded to the python lifetime, but `'static`, as the LUT lives for the entirety of the program.

That's why we need to use `for<'py>` to inform the compiler that the lifetime will be initiated on function call and is not bound to the function itself.

File: py-polars/src/conversion/any_value.rs [hunk 4]
```diff
@@ -124,16 +124,16 @@ pub(crate) fn py_object_to_any_value<'py>(
     strict: bool,
 ) -> PyResult<AnyValue<'py>> {
     // Conversion functions.
-    fn get_null<'py>(_ob: &Bound<'py, PyAny>, _strict: bool) -> PyResult<AnyValue<'py>> {
+    fn get_null<'py>(_ob: &Bound<'py, PyAny>, _strict: bool) -> PyResult<AnyValue<'static>> {
```
Comment (ritchie46) [inline py-polars/src/conversion/any_value.rs:127]: drive bys. All these functions actually return a `'static` as they don't borrow anything.

File: py-polars/src/conversion/any_value.rs (hunk 2 above)
Comment (itamarst) [reply py-polars/src/conversion/any_value.rs:393]: (Unresolving temporarily, just in case someone else reviews)

**Review by ritchie46 — APPROVED** [review]
Alright. Looks great. Thank you @itamarst!

**Pull request merged**